test/cpp/ext/otel/otel_tracing_test.cc

```diff
@@ -0,0 +1,545 @@
+#include "src/core/config/core_configuration.h"
+#include "src/core/telemetry/call_tracer.h"
+#include "src/cpp/ext/otel/otel_plugin.h"
+#include "test/core/test_util/fake_stats_plugin.h"
+#include "test/cpp/end2end/test_service_impl.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+using opentelemetry::sdk::trace::SpanData;
+using opentelemetry::sdk::trace::SpanDataEvent;
+using ::testing::Lt;
+using ::testing::MatchesRegex;
+using ::testing::Pair;
+using ::testing::UnorderedElementsAre;
+using ::testing::VariantWith;
+
+class OTelTracingTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    grpc_core::CoreConfiguration::Reset();
+    grpc_init();
+    data_ =
+        std::make_shared<opentelemetry::exporter::memory::InMemorySpanData>(10);
+    // Register OTel plugin for tracing with an in memory exporter
+    auto tracer_provider =
+        std::make_shared<opentelemetry::sdk::trace::TracerProvider>(
+            opentelemetry::sdk::trace::SimpleSpanProcessorFactory::Create(
+                opentelemetry::exporter::memory::InMemorySpanExporterFactory::
+                    Create(data_)));
+    tracer_ = tracer_provider->GetTracer("grpc-test");
+    ASSERT_TRUE(OpenTelemetryPluginBuilder()
+                    .SetTracerProvider(std::move(tracer_provider))
+                    .SetTextMapPropagator(
+                        grpc::experimental::MakeGrpcTraceBinTextMapPropagator())
+                    .BuildAndRegisterGlobal()
+                    .ok());
+    grpc::ServerBuilder builder;
+    int port;
+    // Use IPv4 here because it's less flaky than IPv6 ("[::]:0") on Travis.
+    builder.AddListeningPort("0.0.0.0:0", grpc::InsecureServerCredentials(),
+                             &port);
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    server_address_ = absl::StrCat("localhost:", port);
+    auto channel = grpc::CreateChannel(server_address_,
+                                       grpc::InsecureChannelCredentials());
+    stub_ = EchoTestService::NewStub(channel);
+  }
+
+  void TearDown() override {
+    server_->Shutdown();
+    grpc_shutdown_blocking();
+    grpc_core::ServerCallTracerFactory::TestOnlyReset();
+    grpc_core::GlobalStatsPluginRegistryTestPeer::
+        ResetGlobalStatsPluginRegistry();
+  }
+
+  void SendRPC() {
+    EchoRequest request;
+    request.set_message("foo");
+    EchoResponse response;
+    grpc::ClientContext context;
+    grpc::Status status = stub_->Echo(&context, request, &response);
+  }
+
+  opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> tracer_;
+  std::shared_ptr<opentelemetry::exporter::memory::InMemorySpanData> data_;
+  CallbackTestServiceImpl service_;
+  std::string server_address_;
+  std::unique_ptr<grpc::Server> server_;
+  std::unique_ptr<EchoTestService::Stub> stub_;
+};
+
+TEST_F(OTelTracingTest, Basic) {
+  SendRPC();
+  absl::SleepFor(absl::Milliseconds(500 * grpc_test_slowdown_factor()));
```

Comment:
I assume these sleeps are needed to let the data propagate into OT?  How do we know this sleep is long enough?